jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -65,78 +65,238 @@ String getName() {
   /** Builds with each layer's files. */
   public static class Builder {
 
-    private final Map<LayerType, List<Path>> layerFilesMap = new EnumMap<>(LayerType.class);
-    private final Map<LayerType, AbsoluteUnixPath> extractionPathMap =
-        new EnumMap<>(LayerType.class);
+    private static class LayerFile {
+      private Path sourceFile;
+      private AbsoluteUnixPath pathInContainer;
+      private boolean recursiveAdd;
+
+      private LayerFile(Path sourceFile, AbsoluteUnixPath pathInContainer, boolean recursiveAdd) {
+        this.sourceFile = sourceFile;
+        this.pathInContainer = pathInContainer;
+        this.recursiveAdd = recursiveAdd;
+      }
+    }
+
+    private final Map<LayerType, List<LayerFile>> layerFilesMap = new EnumMap<>(LayerType.class);
```

Comment:
Could this just be a `Map<LayerType, LayerConfiguration>`? The `add...File` methods would just call `LayerConfiguration#add/addRecursive`.

Reply:
It's certainly doable. But I also think this provides a small degree of convenience to the caller that they don't have to use multiple builders of `LayerConfiguration` for the builder of `JavaLayerConfigurations`. I think if you don't feel strongly against of this API, we could try this for a bit and decide later if we should go for `LayerConfiguration` or think about a totally different API?

Reply:
Oh I wasn't referring to changing the API - just the internal representation here. The API for `JavaLayerConfigurations` could still be the same.

Reply:
Oh, I see. That makes sense. The only reason I am constructing `LayerConfiguration` in `build()` was to maintain the same API where `add...FileRecursive()` methods do not throw an exception. Currently, `LayerConfiguration.Builder.addEntryRecursive()` throws `IOException`, so if we were to change the map to contain `LayerConfiguration`, we will need to expose the `IOException` at the time of calling `add...FileRecursive()`, which may not be bad, since `LayerConfiguration` is already doing so. WDYT?

Reply:
Yea I think having the `IOException` should be fine.